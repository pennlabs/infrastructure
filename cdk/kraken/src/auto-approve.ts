import { Workflow, Stack, WorkflowProps, Job } from "cdkactions";
import { Construct } from "constructs";

/**
 * Auto-approve PRs opened by dependabot. Used to fulfil requirements
 * needed to automatically merge using mergify
 */
export class AutoApproveStack extends Stack {
  /**
   *
   * @param scope cdkactions App instance.
   * @param overrides Optional overrides for the stack.
   */
  public constructor(scope: Construct, overrides?: Partial<WorkflowProps>) {
    super(scope, "autoapprove");
    const workflow = new Workflow(this, "autoapprove", {
      name: "Auto Approve dependabot PRs",
      on: "pullRequest",
      ...overrides,
    });

    new Job(workflow, "approve", {
      runsOn: "ubuntu-latest",
      steps: [
        {
          uses: "hmarr/auto-approve-action@v2.0.0",
          if: "github.actor == 'dependabot[bot]'",
          with: {
            "github-token": "${{ secrets.BOT_GITHUB_PAT }}",
          },
        },
      ],
    });
  }
}
