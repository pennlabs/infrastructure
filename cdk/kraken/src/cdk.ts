import { CheckoutJob, Workflow, Stack, WorkflowProps } from 'cdkactions';
import { Construct } from 'constructs';
import dedent from 'ts-dedent';

/**
 * Optional props to configure the CDK publish stack.
 */
export interface CDKPublishStackProps {
  /**
   * Branch to restrict publishing to
   * @default master
   */
  defaultBranch?: string;

  /**
   * Node version to test the cdk application in
   * @default "14"
   */
  nodeVersion?: string;
}

/**
 * Test, upload code coverage for, and publish a cdk application.
 */
export class CDKPublishStack extends Stack {
  /**
   *
   * @param scope cdkactions App instance.
   * @param id Id of the job.
   * @param config Optional configuration for the cdk publish stack.
   * @param overrides Optional overrides for the stack.
   */
  public constructor(scope: Construct, id: string, config?: CDKPublishStackProps, overrides?: Partial<WorkflowProps>) {
    // Build config
    const fullConfig: Required<CDKPublishStackProps> = {
      defaultBranch: 'master',
      nodeVersion: '14',
      ...config,
    };
    const path = `cdk/${id}`;
    super(scope, id);
    const workflow = new Workflow(this, id, {
      name: `Publish ${id}`,
      on: {
        push: {
          paths: [`${path}/**`],
        },
      },
      ...overrides,
    });
    new CheckoutJob(workflow, 'publish', {
      runsOn: 'ubuntu-latest',
      steps: [{
        name: 'Cache',
        uses: 'actions/cache@v2',
        with: {
          path: '**/node_modules',
          key: `v0-\${{ hashFiles('${path}/yarn.lock') }}`,
        },
      },
      {
        name: 'Install Dependencies',
        run: dedent`cd ${path}
        yarn install --frozen-lockfile`,
      },
      {
        name: 'Test',
        run: dedent`cd ${path}
        yarn test`,
      },
      {
        name: 'Upload Code Coverage',
        run: dedent`ROOT=$(pwd)
        cd ${path}
        yarn run codecov -p $ROOT -F ${id}`,
      },
      {
        name: 'Install jq',
        run: dedent`
        curl -sSLo /usr/bin/jq https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64
        chmod +x /usr/bin/jq`,
      },
      {
        name: 'Check if version is already published to npm',
        run: dedent`cd ${path}
        PACKAGE=\$(node -p "require('./package.json').name")
        VERSION=\$(node -p "require('./package.json').version")
        NEW_VERSION=\$(yarn info $PACKAGE versions --json | jq ".data | any(. == \\"$VERSION\\") | not")
        echo "::set-output name=NEW_VERSION::$NEW_VERSION"`,
      },
      {
        name: 'Publish to npm',
        id: 'version_check',
        run: dedent`cd ${path}
        yarn compile
        yarn package
        mv dist/js/*.tgz dist/js/${id}.tgz
        yarn publish --non-interactive --access public dist/js/${id}.tgz`,
        if: `github.ref == 'refs/heads/${fullConfig.defaultBranch}' && steps.version_check.outputs.NEW_VERSION`,
        env: {
          NPM_AUTH_TOKEN: '${{ secrets.NPM_AUTH_TOKEN }}',
        },
      },
      {
        name: 'Build docs',
        run: dedent`cd ${path}
        yarn docgen`,
      },
      {
        name: 'Publish docs',
        if: `github.ref == 'refs/heads/${fullConfig.defaultBranch}'`,
        uses: 'peaceiris/actions-gh-pages@v3',
        with: {
          personal_token: '${{ secrets.BOT_GITHUB_PAT }}',
          external_repository: `pennlabs/${id}-docs`,
          cname: `${id}.pennlabs.org`,
          publish_branch: 'master',
          publish_dir: `${path}/docs`,
          user_name: 'github-actions',
          user_email: 'github-actions[bot]@users.noreply.github.com',
        },
      }],
      container: {
        image: `node:${fullConfig.nodeVersion}`,
      },
    });
  }
}
