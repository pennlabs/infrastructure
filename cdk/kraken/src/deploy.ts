import { CheckoutJob, Workflow, JobProps } from 'cdkactions';
import dedent from 'ts-dedent';

/**
 * Optional props to configure the deploy job.
 */
export interface DeployJobProps {
  /**
   * Deploy tag to set in kittyhawk.
   * @default current git sha
   */
  deployTag?: string;

  /**
   * Branch to limit deploys to.
   * @default master
   */
  defaultBranch?: string;
}

/**
 * A job to deploy an application using Kittyhawk.
 */
export class DeployJob extends CheckoutJob {
  /**
   *
   * @param scope cdkactions Workflow instance
   * @param config Optional configuration for the deploy job.
   * @param overrides Optinoal overrides for the job.
   */
  public constructor(scope: Workflow, config?: DeployJobProps, overrides?: Partial<JobProps>) {
    // Build config
    const fullConfig: Required<DeployJobProps> = {
      deployTag: '${{ github.sha }}',
      defaultBranch: 'master',
      ...config,
    };

    super(scope, 'deploy', {
      runsOn: 'ubuntu-latest',
      if: `github.ref == 'refs/heads/${fullConfig.defaultBranch}'`,
      steps: [
        {
          id: 'synth',
          name: 'Synth cdk8s manifests',
          run: dedent`cd k8s
          yarn install

          # get repo name (by removing owner/organization)
          RELEASE_NAME=\${REPOSITORY#*/}

          # Export RELEASE_NAME as an output
          echo "::set-output RELEASE_NAME=$RELEASE_NAME"

          yarn build`,
          env: {
            GIT_SHA: fullConfig.deployTag,
            REPOSITORY: '${{ github.repository }}',
            AWS_ACCOUNT_ID: '${{ secrets.AWS_ACCOUNT_ID }}',
          },
        },
        {
          name: 'Deploy',
          run: dedent`aws eks --region us-east-1 update-kubeconfig --name production --role-arn arn:aws:iam::\${AWS_ACCOUNT_ID}:role/kubectl

          # get repo name from synth step
          RELEASE_NAME=\${{ steps.synth.outputs.RELEASE_NAME }}

          # Deploy
          # TODO: figure out labels/deploy command
          kubectl apply -f k8s/dist/ --prune -l app.kubernetes.io/part-of=$RELEASE_NAME`,
          env: {
            AWS_ACCOUNT_ID: '${{ secrets.AWS_ACCOUNT_ID }}',
            AWS_ACCESS_KEY_ID: '${{ secrets.GH_AWS_ACCESS_KEY_ID }}',
            AWS_SECRET_ACCESS_KEY: '${{ secrets.GH_AWS_SECRET_ACCESS_KEY }}',
          },
        },
      ],
      ...overrides,
    });
  }
}
