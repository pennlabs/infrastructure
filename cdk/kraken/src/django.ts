import { CheckoutJob, Workflow, StepsProps, JobProps } from "cdkactions";
import dedent from "ts-dedent";
import { buildId, buildName } from "./utils";

/**
 * Props to configure the Django check job.
 */
export interface DjangoCheckJobProps {
  /**
   * A custom id to append onto job name and ids. Useful when using
   * multiple instances of DjangoCheckJob in a single workflow.
   * @default no suffix
   */
  id?: string;

  /**
   * Python version to test the project with.
   * @default "3.10-bookworm"
   */
  pythonVersion?: string;

  /**
   * Location of the Django project within the repo.
   * @default "."
   */
  path?: string;

  /**
   * Name of the Django project to test.
   */
  projectName: string;

  /**
   * Check the project with black?
   * @default true
   */
  black?: boolean;

  /**
   * Check the project with flake8?
   * @default true
   */
  flake8?: boolean;
}

/**
 * A job to lint and test a Django project as well as upload code coverage.
 *
 * Note that tests are run in parallel to reduce testing time.
 */
export class DjangoCheckJob extends CheckoutJob {
  /**
   *
   * @param scope cdkactions Workflow instance.
   * @param config Configuration for the Django check job.
   * @param overrides Optional overrides for the job.
   */
  public constructor(
    scope: Workflow,
    config: DjangoCheckJobProps,
    overrides?: Partial<JobProps>
  ) {
    // Build config
    const fullConfig: Required<DjangoCheckJobProps> = {
      id: "",
      pythonVersion: "3.10-bookworm",
      path: ".",
      black: true,
      flake8: true,
      ...config,
    };

    // Define steps
    const steps: StepsProps[] = [
      {
        name: "Cache",
        uses: "actions/cache@v4",
        with: {
          path: "~/.local/share/virtualenvs",
          key: `v0-\${{ hashFiles('${fullConfig.path}/Pipfile.lock') }}`,
        },
      },
      {
        name: "Install Dependencies",
        run: dedent`cd ${fullConfig.path}
      pip install pipenv
      pipenv install --deploy --dev`,
      },
    ];
    if (fullConfig.flake8) {
      steps.push({
        name: "Lint (flake8)",
        run: dedent`cd ${fullConfig.path}
        pipenv run flake8 .`,
      });
    }
    if (fullConfig.black) {
      steps.push({
        name: "Lint (black)",
        run: dedent`cd ${fullConfig.path}
        pipenv run black --check .`,
      });
    }
    steps.push({
      name: "Test (run in parallel)",
      run: dedent`cd ${fullConfig.path}
      pipenv run coverage run --concurrency=multiprocessing manage.py test --settings=${fullConfig.projectName}.settings.ci --parallel
      pipenv run coverage combine
      pipenv run coverage xml`,
    });
    steps.push({
      name: "Upload Code Coverage",
      uses: "codecov/codecov-action@v3",
      with: {
        token: "${{ secrets.CODECOV_TOKEN }}",
        directory: "./backend/",
        fail_ci_if_error: true,
        files: "coverage.xml",
        name: "codecov-umbrella",
        verbose: true,
      },
    });

    // Create Job
    super(scope, buildId("django-check", fullConfig.id), {
      name: buildName("Django Check", fullConfig.id),
      runsOn: "ubuntu-latest",
      steps,
      container: {
        image: `python:${fullConfig.pythonVersion}`,
      },
      env: {
        DATABASE_URL: "postgres://postgres:postgres@postgres:5432/postgres",
      },
      services: {
        postgres: {
          image: "postgres:12",
          env: {
            POSTGRES_USER: "postgres",
            POSTGRES_DB: "postgres",
            POSTGRES_PASSWORD: "postgres",
          },
          options:
            "--health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5",
        },
      },
      ...overrides,
    });
  }
}
