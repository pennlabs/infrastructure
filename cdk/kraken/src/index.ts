export * from './cdk';
export * from './deploy';
export * from './django-project';
export * from './django';
export * from './docker';
export * from './integration-tests';
export * from './labs-application';
export * from './pypi';
export * from './react-project';
export * from './react';
export * from './utils';
