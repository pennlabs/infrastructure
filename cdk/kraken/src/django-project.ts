import { JobProps, Workflow } from "cdkactions";
import { DjangoCheckJob, DjangoCheckJobProps } from "./django";
import { DockerPublishJob, DockerPublishJobProps } from "./docker";
import { buildId } from "./utils";

/**
 * Props to configure the DjangoProject.
 */
export interface DjangoProjectProps {
  /**
   * A custom id to append onto job name and ids. Useful when using
   * multiple instances of DjangoProject in a single workflow.
   * @default no suffix
   */
  id?: string;

  /**
   * Name of the Django project to test.
   */
  projectName: string;

  /**
   * Location of the Django project within the repo.
   * @default "."
   */
  path?: string;

  /**
   * Docker image name to publish
   */
  imageName: string;

  /**
   * Optional props to pass to the check job.
   */
  checkProps?: Partial<DjangoCheckJobProps>;

  /**
   * Optional overrides for the check job.
   */
  checkOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the docker publish job.
   */
  publishProps?: Partial<DockerPublishJobProps>;

  /**
   * Optional overrides for the docker publish job.
   */
  publishOverrides?: Partial<JobProps>;
}

/**
 * Creates a DjangoCheckJob and a DockerPublishJob to lint, test, build, and publish
 * a docker image of a django project. The jobs run sequentially.
 */
export class DjangoProject {
  /**
   * ID of the docker publish job.
   */
  public readonly publishJobId: string;

  /**
   * Full name of the docker image built.
   */
  public readonly dockerImageName: string;

  /**
   *
   * @param workflow cdkaction Workflow instance.
   * @param config Configuration for the django project,
   */
  public constructor(workflow: Workflow, config: DjangoProjectProps) {
    // Build config
    const fullConfig: Required<DjangoProjectProps> = {
      id: "",
      path: ".",
      checkProps: {},
      checkOverrides: {},
      publishProps: {},
      publishOverrides: {},
      ...config,
    };

    // Add jobs
    const djangoCheckJob = new DjangoCheckJob(
      workflow,
      {
        id: fullConfig.id,
        projectName: fullConfig.projectName,
        path: fullConfig.path,
        ...config.checkProps,
      },
      fullConfig.checkOverrides
    );

    const publishJob = new DockerPublishJob(
      workflow,
      buildId("backend", fullConfig.id),
      {
        imageName: fullConfig.imageName,
        path: fullConfig.path,
        ...fullConfig.publishProps,
      },
      {
        needs: djangoCheckJob.id,
        ...fullConfig.publishOverrides,
      }
    );

    // Set public fields
    this.publishJobId = publishJob.id;
    this.dockerImageName = publishJob.dockerImageName;
  }
}
