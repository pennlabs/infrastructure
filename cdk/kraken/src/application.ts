import { Workflow, JobProps, WorkflowProps, Stack, CheckoutJob } from 'cdkactions';
import { Construct } from 'constructs';
import { DeployJob, DeployJobProps } from './deploy';
import { DjangoCheckJobProps } from './django';
import { DjangoProject } from './django-project';
import { DockerPublishJobProps } from './docker';
import { ReactCheckJobProps } from './react';
import { ReactProject } from './react-project';


/**
 * Props to configure the Application stack
 */
export interface ApplicationStackProps {
  /**
   * Path to the django project.
   * @default "backend"
   */
  backendPath?: string;

  /**
   * Path to the react project.
   * @default "frontend"
   */
  frontendPath?: string;

  /**
   * If true, run integration tests using docker-compose.
   * @default false
   */
  integrationTests?: boolean;

  /**
   * Base name of the docker images to publish.
   * Will publish $base-frontend and $base-backend images.
   */
  dockerImageBaseName: string;

  /**
   * Name of the django project.
   */
  djangoProjectName: string;

  /**
   * Optional props to pass to the django check job.
   */
  djangoCheckProps?: Partial<DjangoCheckJobProps>;

  /**
   * Optional overrides for the django check job.
   */
  djangoCheckOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the django docker publish job.
   */
  djangoDockerProps?: Partial<DockerPublishJobProps>;

  /**
   * Optional overrides for the django docker publish job.
   */
  djangoDockerOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the react check job.
   */
  reactCheckProps?: Partial<ReactCheckJobProps>;

  /**
   * Optional overrides for the react check job.
   */
  reactCheckOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the react docker publish job.
   */
  reactDockerProps?: Partial<DockerPublishJobProps>;

  /**
   * Optional overrides for the react docker publish job.
   */
  reactDockerOverrides?: Partial<JobProps>;

  /**
   * Optional overrides for the integration tests job.
   */
  integrationOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the deploy job.
   */
  deployProps?: Partial<DeployJobProps>;

  /**
   * Optional overrides for the deploy job.
   */
  deployOverrides?: Partial<JobProps>;
}

/**
 * A stack to test, build, and deploy a fullstack application (Django + React).
 */
export class ApplicationStack extends Stack {
  /**
   *
   * @param scope cdkactions App instance.
   * @param config Configuration for the Application stack.
   * @param overrides Optional overrides for the workflow.
   */
  public constructor(scope: Construct, config: ApplicationStackProps, overrides?: Partial<WorkflowProps>) {
    // Build config
    const fullConfig: Required<ApplicationStackProps> = {
      backendPath: 'backend',
      frontendPath: 'frontend',
      integrationTests: false,
      djangoCheckProps: {},
      djangoCheckOverrides: {},
      djangoDockerProps: {},
      djangoDockerOverrides: {},
      reactCheckProps: {},
      reactCheckOverrides: {},
      reactDockerProps: {},
      reactDockerOverrides: {},
      integrationOverrides: {},
      deployProps: {},
      deployOverrides: {},
      ...config,
    };
    // Create stack
    super(scope, 'application');
    const workflow = new Workflow(this, 'build-and-deploy', {
      name: 'Build and Deploy',
      on: 'push',
      ...overrides,
    });

    // Django
    const djangoProject = new DjangoProject(workflow,
      {
        projectName: fullConfig.djangoProjectName,
        path: fullConfig.backendPath,
        imageName: `${fullConfig.dockerImageBaseName}-backend`,
        checkProps: fullConfig.djangoCheckProps,
        checkOverrides: fullConfig.djangoCheckOverrides,
        publishProps: fullConfig.djangoDockerProps,
        publishOverrides: fullConfig.djangoDockerOverrides,
      });

    // React
    const reactProject = new ReactProject(workflow,
      {
        path: fullConfig.frontendPath,
        imageName: `${fullConfig.dockerImageBaseName}-frontend`,
        checkProps: fullConfig.reactCheckProps,
        checkOverrides: fullConfig.reactCheckOverrides,
        publishProps: fullConfig.reactDockerProps,
        publishOverrides: fullConfig.reactDockerOverrides,
      },
    );

    const integrationTestsId = 'integration-tests';
    if (fullConfig.integrationTests) {
      // TODO: finish this
      new CheckoutJob(workflow, integrationTestsId, {
        name: 'Integration Tests',
        runsOn: 'ubuntu-latest',
        needs: [djangoProject.publishJobId, reactProject.publishJobId],
        steps: [{
          name: 'Run docker compose',
          run: 'docker-compose up -d -f docker-compose.test.yaml',
        },
        {
          name: 'Populate backend',
          run: 'docker run backend python manage.py populate',
        },
        {
          name: 'Run integration tests',
          run: 'docker run frontend yarn integration',
        }],
        env: {
          GIT_SHA: '${{ github.sha }}',
        },
        ...fullConfig.integrationOverrides,
      });
    }

    // Deploy
    const deployNeeds = fullConfig.integrationTests ? [integrationTestsId] : [djangoProject.publishJobId, reactProject.publishJobId];
    new DeployJob(workflow, fullConfig.deployProps,
      {
        needs: deployNeeds,
        ...fullConfig.deployOverrides,
      });
  }
}
