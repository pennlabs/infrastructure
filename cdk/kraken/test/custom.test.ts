import { Workflow } from "cdkactions";
import { DeployJob, DjangoProject, ReactProject } from "../src";

test("workflow with 2 django+react projects", () => {
  const workflow = new Workflow(undefined as any, "workflow", {
    name: "Workflow",
    on: "push",
  });

  const djangoOne = new DjangoProject(workflow, {
    id: "one",
    projectName: "projectOne",
    path: "backendOne",
    imageName: "imageOne-backend",
  });
  const djangoTwo = new DjangoProject(workflow, {
    id: "two",
    projectName: "projectTwo",
    path: "backendTwo",
    imageName: "imageTwo-backend",
  });

  const reactOne = new ReactProject(workflow, {
    id: "one",
    path: "frontendOne",
    imageName: "imageOne-frontend",
  });
  const reactTwo = new ReactProject(workflow, {
    id: "two",
    path: "frontendTwo",
    imageName: "imageTwo-frontend",
  });
  new DeployJob(
    workflow,
    {},
    {
      needs: [
        djangoOne.publishJobId,
        djangoTwo.publishJobId,
        reactOne.publishJobId,
        reactTwo.publishJobId,
      ],
    }
  );

  expect(workflow.toGHAction()).toMatchSnapshot();
});
