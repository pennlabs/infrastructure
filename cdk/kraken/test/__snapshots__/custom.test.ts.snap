// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workflow with 2 django+react projects 1`] = `
Object {
  "jobs": Object {
    "deploy": Object {
      "if": "github.ref == 'refs/heads/master'",
      "needs": Array [
        "publish-backend-one",
        "publish-backend-two",
        "publish-frontend-one",
        "publish-frontend-two",
      ],
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "env": Object {
            "AWS_ACCOUNT_ID": "\${{ secrets.AWS_ACCOUNT_ID }}",
            "GIT_SHA": "\${{ github.sha }}",
            "REPOSITORY": "\${{ github.repository }}",
          },
          "id": "synth",
          "name": "Synth cdk8s manifests",
          "run": "cd k8s
yarn install

# get repo name (by removing owner/organization)
RELEASE_NAME=\${REPOSITORY#*/}

# Export RELEASE_NAME as an output
echo \\"::set-output RELEASE_NAME=$RELEASE_NAME\\"

yarn build",
        },
        Object {
          "env": Object {
            "AWS_ACCESS_KEY_ID": "\${{ secrets.GH_AWS_ACCESS_KEY_ID }}",
            "AWS_ACCOUNT_ID": "\${{ secrets.AWS_ACCOUNT_ID }}",
            "AWS_SECRET_ACCESS_KEY": "\${{ secrets.GH_AWS_SECRET_ACCESS_KEY }}",
          },
          "name": "Deploy",
          "run": "aws eks --region us-east-1 update-kubeconfig --name production --role-arn arn:aws:iam::\${AWS_ACCOUNT_ID}:role/kubectl

# get repo name from synth step
RELEASE_NAME=\${{ steps.synth.outputs.RELEASE_NAME }}

# Deploy
# TODO: figure out labels/deploy command
kubectl apply -f k8s/dist/ --prune -l app.kubernetes.io/part-of=$RELEASE_NAME",
        },
      ],
    },
    "django-check-one": Object {
      "container": Object {
        "image": "python:3.8-buster",
      },
      "env": Object {
        "DATABASE_URL": "postgres://postgres:postgres@postgres:5432/postgres",
      },
      "name": "Django Check one",
      "runs-on": "ubuntu-latest",
      "services": Object {
        "postgres": Object {
          "env": Object {
            "POSTGRES_DB": "postgres",
            "POSTGRES_PASSWORD": "postgres",
            "POSTGRES_USER": "postgres",
          },
          "image": "postgres:12",
          "options": "--health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5",
        },
      },
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "name": "Cache",
          "uses": "actions/cache@v2",
          "with": Object {
            "key": "v0-\${{ hashFiles('backendOne/Pipfile.lock') }}",
            "path": "~/.local/share/virtualenvs",
          },
        },
        Object {
          "name": "Install Dependencies",
          "run": "cd backendOne
pip install pipenv
pipenv install --deploy --dev",
        },
        Object {
          "name": "Lint (flake8)",
          "run": "cd backendOne
pipenv run flake8 .",
        },
        Object {
          "name": "Lint (black)",
          "run": "cd backendOne
pipenv run black --check .",
        },
        Object {
          "name": "Test (run in parallel)",
          "run": "cd backendOne
pipenv run coverage run --concurrency=multiprocessing manage.py test --settings=projectOne.settings.ci --parallel
pipenv run coverage combine",
        },
        Object {
          "name": "Upload Code Coverage",
          "run": "ROOT=$(pwd)
cd backendOne
pipenv run codecov --root $ROOT --flags backend",
        },
      ],
    },
    "django-check-two": Object {
      "container": Object {
        "image": "python:3.8-buster",
      },
      "env": Object {
        "DATABASE_URL": "postgres://postgres:postgres@postgres:5432/postgres",
      },
      "name": "Django Check two",
      "runs-on": "ubuntu-latest",
      "services": Object {
        "postgres": Object {
          "env": Object {
            "POSTGRES_DB": "postgres",
            "POSTGRES_PASSWORD": "postgres",
            "POSTGRES_USER": "postgres",
          },
          "image": "postgres:12",
          "options": "--health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5",
        },
      },
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "name": "Cache",
          "uses": "actions/cache@v2",
          "with": Object {
            "key": "v0-\${{ hashFiles('backendTwo/Pipfile.lock') }}",
            "path": "~/.local/share/virtualenvs",
          },
        },
        Object {
          "name": "Install Dependencies",
          "run": "cd backendTwo
pip install pipenv
pipenv install --deploy --dev",
        },
        Object {
          "name": "Lint (flake8)",
          "run": "cd backendTwo
pipenv run flake8 .",
        },
        Object {
          "name": "Lint (black)",
          "run": "cd backendTwo
pipenv run black --check .",
        },
        Object {
          "name": "Test (run in parallel)",
          "run": "cd backendTwo
pipenv run coverage run --concurrency=multiprocessing manage.py test --settings=projectTwo.settings.ci --parallel
pipenv run coverage combine",
        },
        Object {
          "name": "Upload Code Coverage",
          "run": "ROOT=$(pwd)
cd backendTwo
pipenv run codecov --root $ROOT --flags backend",
        },
      ],
    },
    "publish-backend-one": Object {
      "name": "Publish backend-one",
      "needs": "django-check-one",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "uses": "docker/setup-qemu-action@v1",
        },
        Object {
          "uses": "docker/setup-buildx-action@v1",
        },
        Object {
          "name": "Cache Docker layers",
          "uses": "actions/cache@v2",
          "with": Object {
            "key": "buildx-publish-backend-one",
            "path": "/tmp/.buildx-cache",
          },
        },
        Object {
          "uses": "docker/login-action@v1",
          "with": Object {
            "password": "\${{ secrets.DOCKER_PASSWORD }}",
            "username": "\${{ secrets.DOCKER_USERNAME }}",
          },
        },
        Object {
          "name": "Build/Publish",
          "uses": "docker/build-push-action@v2",
          "with": Object {
            "cache-from": "type=local,src=/tmp/.buildx-cache,type=registry,ref=pennlabs/imageOne-backend:latest",
            "cache-to": "type=local,dest=/tmp/.buildx-cache",
            "context": "backendOne",
            "file": "backendOne/Dockerfile",
            "push": "\${{ github.ref == 'refs/heads/master' }}",
            "tags": "pennlabs/imageOne-backend:latest,pennlabs/imageOne-backend:\${{ github.sha }}",
          },
        },
      ],
    },
    "publish-backend-two": Object {
      "name": "Publish backend-two",
      "needs": "django-check-two",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "uses": "docker/setup-qemu-action@v1",
        },
        Object {
          "uses": "docker/setup-buildx-action@v1",
        },
        Object {
          "name": "Cache Docker layers",
          "uses": "actions/cache@v2",
          "with": Object {
            "key": "buildx-publish-backend-two",
            "path": "/tmp/.buildx-cache",
          },
        },
        Object {
          "uses": "docker/login-action@v1",
          "with": Object {
            "password": "\${{ secrets.DOCKER_PASSWORD }}",
            "username": "\${{ secrets.DOCKER_USERNAME }}",
          },
        },
        Object {
          "name": "Build/Publish",
          "uses": "docker/build-push-action@v2",
          "with": Object {
            "cache-from": "type=local,src=/tmp/.buildx-cache,type=registry,ref=pennlabs/imageTwo-backend:latest",
            "cache-to": "type=local,dest=/tmp/.buildx-cache",
            "context": "backendTwo",
            "file": "backendTwo/Dockerfile",
            "push": "\${{ github.ref == 'refs/heads/master' }}",
            "tags": "pennlabs/imageTwo-backend:latest,pennlabs/imageTwo-backend:\${{ github.sha }}",
          },
        },
      ],
    },
    "publish-frontend-one": Object {
      "name": "Publish frontend-one",
      "needs": "react-check-one",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "uses": "docker/setup-qemu-action@v1",
        },
        Object {
          "uses": "docker/setup-buildx-action@v1",
        },
        Object {
          "name": "Cache Docker layers",
          "uses": "actions/cache@v2",
          "with": Object {
            "key": "buildx-publish-frontend-one",
            "path": "/tmp/.buildx-cache",
          },
        },
        Object {
          "uses": "docker/login-action@v1",
          "with": Object {
            "password": "\${{ secrets.DOCKER_PASSWORD }}",
            "username": "\${{ secrets.DOCKER_USERNAME }}",
          },
        },
        Object {
          "name": "Build/Publish",
          "uses": "docker/build-push-action@v2",
          "with": Object {
            "cache-from": "type=local,src=/tmp/.buildx-cache,type=registry,ref=pennlabs/imageOne-frontend:latest",
            "cache-to": "type=local,dest=/tmp/.buildx-cache",
            "context": "frontendOne",
            "file": "frontendOne/Dockerfile",
            "push": "\${{ github.ref == 'refs/heads/master' }}",
            "tags": "pennlabs/imageOne-frontend:latest,pennlabs/imageOne-frontend:\${{ github.sha }}",
          },
        },
      ],
    },
    "publish-frontend-two": Object {
      "name": "Publish frontend-two",
      "needs": "react-check-two",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "uses": "docker/setup-qemu-action@v1",
        },
        Object {
          "uses": "docker/setup-buildx-action@v1",
        },
        Object {
          "name": "Cache Docker layers",
          "uses": "actions/cache@v2",
          "with": Object {
            "key": "buildx-publish-frontend-two",
            "path": "/tmp/.buildx-cache",
          },
        },
        Object {
          "uses": "docker/login-action@v1",
          "with": Object {
            "password": "\${{ secrets.DOCKER_PASSWORD }}",
            "username": "\${{ secrets.DOCKER_USERNAME }}",
          },
        },
        Object {
          "name": "Build/Publish",
          "uses": "docker/build-push-action@v2",
          "with": Object {
            "cache-from": "type=local,src=/tmp/.buildx-cache,type=registry,ref=pennlabs/imageTwo-frontend:latest",
            "cache-to": "type=local,dest=/tmp/.buildx-cache",
            "context": "frontendTwo",
            "file": "frontendTwo/Dockerfile",
            "push": "\${{ github.ref == 'refs/heads/master' }}",
            "tags": "pennlabs/imageTwo-frontend:latest,pennlabs/imageTwo-frontend:\${{ github.sha }}",
          },
        },
      ],
    },
    "react-check-one": Object {
      "container": Object {
        "image": "node:14",
      },
      "name": "React Check one",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "name": "Cache",
          "uses": "actions/cache@v2",
          "with": Object {
            "key": "v0-\${{ hashFiles('frontendOne/yarn.lock') }}",
            "path": "**/node_modules",
          },
        },
        Object {
          "name": "Install Dependencies",
          "run": "cd frontendOne
yarn install --frozen-lockfile",
        },
        Object {
          "name": "Lint",
          "run": "cd frontendOne
yarn lint",
        },
        Object {
          "name": "Test",
          "run": "cd frontendOne
yarn test",
        },
        Object {
          "name": "Upload Code Coverage",
          "run": "ROOT=$(pwd)
cd frontendOne
yarn run codecov -p $ROOT -F frontend",
        },
      ],
    },
    "react-check-two": Object {
      "container": Object {
        "image": "node:14",
      },
      "name": "React Check two",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "name": "Cache",
          "uses": "actions/cache@v2",
          "with": Object {
            "key": "v0-\${{ hashFiles('frontendTwo/yarn.lock') }}",
            "path": "**/node_modules",
          },
        },
        Object {
          "name": "Install Dependencies",
          "run": "cd frontendTwo
yarn install --frozen-lockfile",
        },
        Object {
          "name": "Lint",
          "run": "cd frontendTwo
yarn lint",
        },
        Object {
          "name": "Test",
          "run": "cd frontendTwo
yarn test",
        },
        Object {
          "name": "Upload Code Coverage",
          "run": "ROOT=$(pwd)
cd frontendTwo
yarn run codecov -p $ROOT -F frontend",
        },
      ],
    },
  },
  "name": "Workflow",
  "on": "push",
}
`;
